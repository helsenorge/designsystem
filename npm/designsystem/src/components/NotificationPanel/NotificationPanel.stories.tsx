import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import NotificationPanel from './NotificationPanel';
import { allTitleTags } from '../../../.storybook/knobs';
import Docs from '../../docs';
import { getColor } from '../../theme/currys';

const meta = {
  title: '@helsenorge/designsystem-react/Components/NotificationPanel',
  component: NotificationPanel,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={NotificationPanel} />,
      description: {
        component: 'Et NotificationPanel lar deg vise viktig informasjon avskilt fra bakgrunnen med ulike farger og ikoner',
      },
    },
  },
  args: {
    dismissable: false,
    size: 'large',
    fluid: false,
    label: 'Det har skjedd noe galt. Prøv igjen senere.',
    labelHtmlMarkup: 'h1',
    variant: 'error',
    role: undefined,
  },
  argTypes: {
    dismissable: {
      control: 'boolean',
    },
    size: {
      control: 'select',
      options: ['small', 'medium', 'large'],
    },
    fluid: {
      control: 'boolean',
    },
    label: {
      control: 'text',
    },
    labelHtmlMarkup: {
      control: 'select',
      options: allTitleTags,
    },
    variant: {
      control: 'select',
      options: ['info', 'warn', 'error', 'success'],
    },
    role: {
      control: 'select',
      options: ['', 'alert', 'region'],
    },
  },
} satisfies Meta<typeof NotificationPanel>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => <NotificationPanel {...args} label="Label only"></NotificationPanel>,
};

export const WithoutLabel: Story = {
  render: args => (
    <>
      <NotificationPanel {...args}>
        <span>{'Eksempel med label'}</span>
      </NotificationPanel>
      <br />
      <NotificationPanel {...args} label={undefined}>
        <span>{'Eksempel uten label. Nå får innhold en font-vekt som ligner mer på label'}</span>
      </NotificationPanel>
      <br />
      <NotificationPanel
        {...args}
        label={undefined}
        expanderChildren={'Dette er innhold i expanderen. Det vises først når man trykker på knappen.'}
        expanderButtonText="Vis mindre"
        expanderButtonClosedText="Vis mer"
      >
        <span>{'Eksempel uten label. Nå får innhold en font-vekt som ligner mer på label'}</span>
      </NotificationPanel>
    </>
  ),
};

export const AllVariants: Story = {
  render: args => (
    <>
      <div className="row">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'info'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere.`}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'warn'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere. `}
              <a href={'https://www.helsenorge.no'} target="_blank" rel="noopener noreferrer">
                {'Les mer om dine rettigheter her (external).'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'error'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere. `}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} variant={'success'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere. `}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
    </>
  ),
};

export const Expander: Story = {
  render: args => (
    <>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel
            {...args}
            label={'Med expander'}
            expanderChildren={'Dette er innhold i expanderen. Det vises først når man trykker på knappen.'}
            expanderButtonText="Vis mindre"
            expanderButtonClosedText="Vis mer"
          >
            {'Dette er innhold'}
          </NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel
            {...args}
            label={'Med expander'}
            expanderChildren={'Dette er innhold i expanderen. Dette innholdet er åpnet fra start'}
            expanderButtonText="Vis mindre"
            expanderButtonClosedText="Vis mer"
            expanderOpenFromStart
          >
            {'Dette er innhold'}
          </NotificationPanel>
        </div>
      </div>
    </>
  ),
};
export const Dismissable: Story = {
  render: args => (
    <NotificationPanel {...args} label={'Dismissable'} dismissable>
      {'Dette er dismissesable'}
    </NotificationPanel>
  ),
};

export const Compact: Story = {
  render: args => (
    <div style={{ backgroundColor: getColor('blueberry', 50), padding: '3rem' }}>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={'Compact - basic'} compactVariant={'basic'}></NotificationPanel>{' '}
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={'Compact - outline'} compactVariant={'outline'}></NotificationPanel>
        </div>
      </div>
    </div>
  ),
};

export const CompactWithChildren: Story = {
  render: args => (
    <div style={{ backgroundColor: getColor('blueberry', 50), padding: '3rem' }}>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={undefined} compactVariant={'basic'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere.`}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>{' '}
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} label={undefined} compactVariant={'outline'}>
            <div>
              {`På grunn av kommunesammenslåingen ved nyttår vil enkelte Helsenorge-tjenester være ustabile eller tidsvis utilgjenglig i en
              periode i begynnelsen av januar. Vi beklager ulempnene dette medfører og oppfordrer til å prøve igjen senere.`}
              <a href={'https://www.helsenorge.no'} target="_parent">
                {'Les mer om dine rettigheter her.'}
              </a>
            </div>
          </NotificationPanel>
        </div>
      </div>
    </div>
  ),
};

export const WithSetWidth: Story = {
  render: args => (
    <>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} size="small" label={'Size: small'}></NotificationPanel>{' '}
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} size="medium" label={'Size: medium'}></NotificationPanel>
        </div>
      </div>
      <div className="row mt-6">
        <div className={'col-12'}>
          <NotificationPanel {...args} size="large" label={'Size: large'}></NotificationPanel>
        </div>
      </div>
    </>
  ),
};
