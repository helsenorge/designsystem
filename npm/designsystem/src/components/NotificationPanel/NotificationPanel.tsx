import React from 'react';

import classNames from 'classnames';

import { DetailButton } from './DetailButton/DetailButton';
import { AnalyticsId, IconSize } from '../../constants';
import { useUuid } from '../../hooks/useUuid';
import { getColor } from '../../theme/currys';
import { palette } from '../../theme/palette';
import { getAriaLabelAttributes } from '../../utils/accessibility';
import Close from '../Close';
import Icon from '../Icons';
import CheckFill from '../Icons/CheckFill';
import ErrorSignFill from '../Icons/ErrorSignFill';
import InfoSignFill from '../Icons/InfoSignFill';
import TriangleX from '../Icons/TriangleX';

import styles from './styles.module.scss';

export type NotificationPanelVariants = 'info' | 'warn' | 'alert' | 'success';
export type NotificationCompactVariants = 'basic' | 'outline';
export type NotificationPanelSizes = 'small' | 'medium' | 'large';

export interface NotificationPanelProps {
  /** Adds custom classes to the element. */
  className?: string;
  /** Adds inner child elements. */
  children?: React.ReactNode;
  /** Adds inner expander elements. */
  expanderChildren?: React.ReactNode;
  /** Text for expanderButton. */
  expanderButtonText?: string;
  /** Text for expanderButton when closed. */
  expanderButtonClosedText?: string;
  /** Makes expander be open from start. */
  expanderOpenFromStart?: boolean;
  /** Changes the visual representation of the notification panel. */
  variant?: NotificationPanelVariants;
  /** Makes the panel more compact. Available in basic and outline. */
  compactVariant?: NotificationCompactVariants;
  /** Sets a fixed size for the content container. */
  size?: NotificationPanelSizes;
  /** Used in combination with dismissiable property to close the notification panel. */
  onClick?: (e?: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  /** Toggles the close button in the top right corner. */
  dismissable?: boolean;
  /** Enables a fluid outer container that spans the entire width of parent. */
  fluid?: boolean;
  /** Sets a label for the notification panel. */
  label?: string;
  /** Close button aria-label */
  ariaLabelCloseBtn?: string;
  /** Custom id for the label */
  labelId?: string;
  /** Custom role for the panel. Default is no role. If variant is alert or crisis, the aria role will be set to "alert" unless the role-prop is also set. */
  role?: 'region' | 'alert';
  /** Sets the data-testid attribute. */
  testId?: string;
}

type WrapFluidProps = Pick<NotificationPanelProps, 'fluid' | 'variant'> & {
  children: React.ReactElement;
};

const FluidWrapper: React.FC<WrapFluidProps> = ({ fluid, variant, children }) => {
  if (fluid) {
    const fluidClasses = classNames(styles['fluid-wrapper'], { [styles[`fluid-wrapper--${variant}`]]: variant });

    return <div className={fluidClasses}>{children}</div>;
  }
  return children;
};

const NotificationPanel = React.forwardRef<HTMLDivElement, NotificationPanelProps>((props, ref) => {
  const {
    children,
    variant = 'info',
    dismissable = false,
    onClick,
    expanderChildren,
    expanderButtonText,
    expanderButtonClosedText,
    expanderOpenFromStart = false,
    compactVariant,
    label,
    fluid = false,
    size,
    className,
    labelId,
    role,
    testId,
  } = props;
  const uuid = useUuid(labelId);
  const variantToIconMap = {
    info: (
      <Icon
        svgIcon={InfoSignFill}
        color={palette.blueberry700}
        hoverColor={palette.blueberry700}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
    warn: (
      <Icon
        svgIcon={ErrorSignFill}
        color={palette.banana700}
        hoverColor={palette.banana700}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
    alert: (
      <Icon
        svgIcon={TriangleX}
        color={palette.cherry700}
        hoverColor={palette.cherry700}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
    success: (
      <Icon
        svgIcon={CheckFill}
        color={palette.kiwi900}
        hoverColor={palette.kiwi900}
        size={compactVariant ? IconSize.XSmall : IconSize.Small}
      />
    ),
  };
  const renderContent = (): JSX.Element => {
    const contentClasses = classNames(styles['notification-panel__content']);
    const labelClasses = classNames(styles['notification-panel__label'], {
      [styles['notification-panel__label--no-content']]: !children && !expanderChildren,
      [styles['notification-panel__label__compact']]: !!compactVariant,
      [styles['notification-panel__label__compact--basic']]: compactVariant === 'basic',
    });
    const childrenClasses = classNames(styles['notification-panel__children'], {
      [styles['notification-panel__label-and-content--spacing']]: label,
    });

    return (
      <div className={contentClasses} id={!label ? uuid : undefined}>
        {label && (
          <h1 className={labelClasses} id={uuid}>
            {label}
          </h1>
        )}
        {children && !compactVariant && <div className={childrenClasses}>{children}</div>}
        {expanderChildren && expanderButtonText && expanderButtonClosedText && !compactVariant && (
          <DetailButton
            expanderOpenFromStart={expanderOpenFromStart}
            content={expanderChildren}
            buttonText={expanderButtonText}
            buttonClosedText={expanderButtonClosedText}
          />
        )}
      </div>
    );
  };

  const notificationPanelClasses = classNames(
    styles['notification-panel'],
    styles[`notification-panel--${variant}`],
    {
      [styles[`notification-panel--${size}`]]: !!size,
      [styles['notification-panel__compact']]: !!compactVariant,
      [styles['notification-panel__compact--basic']]: compactVariant === 'basic',
      [styles['notification-panel__compact--outline']]: compactVariant === 'outline',
      [styles['notification-panel--has-children']]: !!children,
      [styles['notification-panel--with-content']]: expanderChildren || (label && children),
      [styles['notification-panel--dismissable']]: dismissable,
    },
    className
  );

  const ariaRole = role || (variant === 'alert' && 'alert') || undefined;
  const ariaLabelAttributes = ariaRole ? getAriaLabelAttributes({ label, id: uuid }) : undefined;

  return (
    <FluidWrapper fluid={fluid} variant={variant}>
      <div
        ref={ref}
        role={ariaRole}
        data-testid={testId}
        data-analyticsid={AnalyticsId.NotificationPanel}
        className={notificationPanelClasses}
        {...ariaLabelAttributes}
      >
        <span className={styles['notification-panel__icon']}>{variantToIconMap[variant]}</span>
        {dismissable && (
          <span className={styles['notification-panel__close']}>
            <Close ariaLabel={props.ariaLabelCloseBtn} onClick={onClick} color={getColor('black')} />
          </span>
        )}
        {renderContent()}
      </div>
    </FluidWrapper>
  );
});

export default NotificationPanel;
