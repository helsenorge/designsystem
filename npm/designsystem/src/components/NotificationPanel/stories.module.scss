@import '../../scss/_breakpoints.scss';
.container {
  margin-left: auto;
  margin-right: auto;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  width: 100%;
  display: flex;

  @media (min-width: map-get($grid-breakpoints, md)) {
    max-width: 66rem;
  }

  @media (min-width: map-get($grid-breakpoints, xl)) {
    max-width: 84rem;
  }
}
