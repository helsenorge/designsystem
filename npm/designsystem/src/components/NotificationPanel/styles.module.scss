@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/font-settings';
@import '../../scss/palette';
@import '../AnchorLink/styles.module';

.notification-panel {
  $panel: &;

  display: flex;
  padding: getSpacer(2xs) getSpacer(s) getSpacer(2xs) getSpacer(2xs);
  box-shadow: 0.25rem 0.25rem 0.375rem rgb(0 0 0 / 15%);
  border-radius: 0.25rem;

  @media (min-width: map.get($grid-breakpoints, md)) {
    padding-left: getSpacer(s);
    padding-right: getSpacer(m);
    #{$panel}__icon {
      margin-right: getSpacer(m);
    }
  }

  &--compact {
    width: fit-content;
    font-size: $font-size-xs;
    line-height: 1.25rem;
    box-shadow: none;

    @media (min-width: map.get($grid-breakpoints, md)) {
      padding-left: getSpacer(2xs);
      padding-right: getSpacer(s);
    }

    &--outline {
      border: none;
    }
  }
  #{$panel}__icon {
    margin-right: getSpacer(2xs);
  }

  &--dismissable {
    padding: getSpacer(2xs) getSpacer(2xs) getSpacer(m) getSpacer(2xs);
  }

  &--small {
    max-width: map.get($grid-breakpoints, sm);
  }

  &--medium {
    max-width: map.get($grid-breakpoints, md);
  }

  &--large {
    max-width: map.get($grid-breakpoints, lg);
  }

  &--info {
    border: 0.125rem solid $blueberry700;
    background-color: $blueberry50;
  }

  &--info#{&}--compact {
    box-shadow: none;
    background-color: transparent;
  }

  &--info#{&}--compact--basic {
    border: none;
  }

  &--warn {
    border: 0.125rem solid $banana700;
    background-color: $banana50;
  }

  &--warn#{&}--compact {
    box-shadow: none;
    background-color: transparent;
  }

  &--warn#{&}--compact--basic {
    border: none;
  }

  &--alert {
    border: 0.125rem solid $cherry700;
    background-color: $cherry50;
  }

  &--alert#{&}--compact {
    box-shadow: none;
    background-color: transparent;
  }

  &--alert#{&}--compact--basic {
    border: none;
  }

  &--success {
    border: 0.125rem solid $kiwi900;
    background-color: $kiwi50;
  }

  &--success#{&}--compact {
    box-shadow: none;
    background-color: transparent;
  }

  &--success#{&}--compact--basic {
    border: none;
  }

  &__close {
    order: 1;
    margin-left: auto;
    padding-left: getSpacer(2xs);
  }

  &__content {
    align-self: center;

    a {
      @include anchorlink;
      @include anchorlink-icon;
    }
  }

  &__label,
  &__children {
    font-size: $font-size-xs;
    line-height: $lineheight-size-xs;
    margin: 0;

    @media (min-width: map.get($grid-breakpoints, md)) {
      font-size: $font-size-sm;
      line-height: 1.625rem;
    }
  }

  &__label {
    font-weight: 700;
    margin: getSpacer(2xs) 0 0 0;

    &--compact {
      font-size: $font-size-xs;
      line-height: 1.25rem;
    }

    &--no-content {
      margin: 0;
    }

    &--compact,
    &--no-content {
      font-weight: 600;
    }

    &--outline {
      font-weight: 500;
    }
  }

  &__children {
    font-weight: 600;
    padding: getSpacer(3xs) 0;

    &--with-label {
      font-weight: 500;
    }

    &--expander-no-label {
      margin: getSpacer(2xs) 0 0 0;

      @media (max-width: map.get($grid-breakpoints, lg)) {
        padding-top: 0;
      }
    }

    &--outline {
      font-weight: 500;
    }
  }

  &__icon {
    line-height: 0;
  }

  &--with-content {
    padding-bottom: getSpacer(xs);
  }
}

.fluid-wrapper {
  display: flex;
  justify-content: center;
  width: 100%;
}
