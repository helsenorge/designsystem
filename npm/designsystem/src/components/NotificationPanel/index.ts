import NotificationPanel from './NotificationPanel';
export * from './NotificationPanel';
export default NotificationPanel;
