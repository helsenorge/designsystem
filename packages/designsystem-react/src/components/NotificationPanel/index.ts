import NotificationPanel from './NotificationPanel';
export default NotificationPanel;
