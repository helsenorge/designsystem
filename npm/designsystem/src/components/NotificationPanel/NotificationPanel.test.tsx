import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import NotificationPanel from './NotificationPanel';

describe('Gitt at NotificationPanel skal vises', () => {
  describe('Når panelet vises med default props', () => {
    test('Så rendres komponenten riktig', (): void => {
      const { container } = render(<NotificationPanel>Some text here for testing.</NotificationPanel>);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når panelet skal kunne lukkes', () => {
    test('Så kalles onClick-handler når man klikker på lukkeknappen', async (): Promise<void> => {
      const mockClose = jest.fn();

      render(
        <NotificationPanel onClick={mockClose} dismissable ariaLabelCloseBtn="Lukk denne beskjeden">
          Some text here for testing.
        </NotificationPanel>
      );

      const closeButton = screen.getByRole('button', { name: 'Lukk denne beskjeden' });
      await userEvent.click(closeButton);
      expect(mockClose).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når panelet kan lukkes', () => {
    test('Så kalles onClick-handleren når man klikker på knappen', async (): Promise<void> => {
      const mockOnClick = jest.fn();
      const { container } = render(
        <NotificationPanel dismissable onClick={mockOnClick}>
          Some text here for testing.
        </NotificationPanel>
      );
      expect(container).toMatchSnapshot();

      const closeButton = screen.getByRole('button');
      await userEvent.click(closeButton);

      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når testId-prop er satt', (): void => {
    test('Så kan komponenten finnes ved hjelp av testId', (): void => {
      render(<NotificationPanel testId="bare-tester">Some text here for testing.</NotificationPanel>);

      const component = screen.getByTestId('bare-tester');
      expect(component).toBeVisible();
    });
  });

  describe('Når bruker klikker på tab', (): void => {
    test('Så får lukkeknappen fokus først', async (): Promise<void> => {
      render(
        <NotificationPanel dismissable>
          Some text here for testing.<a href="/">Lenke</a>
        </NotificationPanel>
      );

      await userEvent.tab();

      const closeButton = screen.getByRole('button', { name: 'Lukk' });

      expect(closeButton).toHaveFocus();
    });
  });

  describe('Når bruker klikker på expand', (): void => {
    test('Så vises innholdet i expander', async (): Promise<void> => {
      render(
        <NotificationPanel expanderButtonClosedText="Close" expanderButtonText="Open" expanderChildren="Expanded Content">
          Some text here for testing.<a href="/">Lenke</a>
        </NotificationPanel>
      );

      const expandButton = screen.getByTestId('expand');
      await userEvent.click(expandButton);
      expect(await screen.findByText('Expanded Content')).toBeInTheDocument();
    });
  });

  describe('Når bruker ser compact-variant', (): void => {
    test('Så vises innhold', async (): Promise<void> => {
      const { container } = render(
        <>
          <NotificationPanel compactVariant="basic">
            Some text here for testing.<a href="/">Lenke</a>
          </NotificationPanel>
          <NotificationPanel compactVariant="outline">
            Some text here for testing.<a href="/">Lenke</a>
          </NotificationPanel>
        </>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når role er region', (): void => {
    describe('Når panelet har label og tekstlig innhold', (): void => {
      test('Så er panelet tilgjengelig via label', (): void => {
        render(
          <NotificationPanel role="region" label="Overskrift">
            Some text here for testing
          </NotificationPanel>
        );

        const panel = screen.getByRole('region', { name: 'Overskrift' });
        expect(panel).toBeInTheDocument();
      });
    });

    describe('Når panelet har bare rent tekstlig innhold', (): void => {
      test('Så er panelet tilgjengelig via teksten i panelet', (): void => {
        render(<NotificationPanel role="region">Some text here for testing</NotificationPanel>);

        const panel = screen.getByRole('region', { name: 'Some text here for testing' });
        expect(panel).toBeInTheDocument();
      });
    });

    describe('Når panelet JSX som innhold, og ikke ariaLabel eller label', (): void => {
      test('Så er panelet tilgjengelig via teksten i panelet', (): void => {
        render(
          <NotificationPanel role="region">
            <p>Some text here for testing</p>
          </NotificationPanel>
        );

        const panel = screen.getByRole('region', { name: 'Some text here for testing' });
        expect(panel).toBeInTheDocument();
      });
    });

    describe('Når panelet har label og labelId', (): void => {
      test('Så er panelet tilgjengelig via label', (): void => {
        render(
          <NotificationPanel role="region" label="Overskrift" labelId={'sidetittel'}>
            Some text here for testing
          </NotificationPanel>
        );

        const panel = screen.getByRole('region', { name: 'Overskrift' });
        expect(panel).toBeInTheDocument();
      });
      test('Så bruker label labelId som id', (): void => {
        render(
          <NotificationPanel role="region" label="Overskrift" labelId={'sidetittel'}>
            Some text here for testing
          </NotificationPanel>
        );

        const title = screen.getByRole('heading', { name: 'Overskrift' });
        expect(title).toHaveAttribute('id', 'sidetittel');
      });
    });
  });

  describe('Når role er alert', (): void => {
    test('Så har komponenten role=alert', (): void => {
      render(<NotificationPanel role="alert">Warning!</NotificationPanel>);

      const component = screen.getByRole('alert', { name: 'Warning!' });
      expect(component).toBeVisible();
    });
  });

  describe('Når role ikke har satt', (): void => {
    test('Så har komponenten ikke role=region', (): void => {
      render(<NotificationPanel>Important!</NotificationPanel>);

      const component = screen.queryByRole('region');
      expect(component).not.toBeInTheDocument();
    });
  });

  describe('Når variant er alert', (): void => {
    test('Så har komponenten role=alert', (): void => {
      render(<NotificationPanel variant="alert">Danger!</NotificationPanel>);

      const component = screen.getByRole('alert', { name: 'Danger!' });
      expect(component).toBeVisible();
    });
  });

  describe('Når variant er alert og role er region', (): void => {
    test('Så har komponenten role=region', (): void => {
      render(
        <NotificationPanel variant="alert" role="region">
          Danger!
        </NotificationPanel>
      );

      const component = screen.getByRole('region', { name: 'Danger!' });
      expect(component).toBeVisible();
    });
  });
});
