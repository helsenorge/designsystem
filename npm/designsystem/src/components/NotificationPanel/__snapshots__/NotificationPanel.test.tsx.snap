// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at NotificationPanel skal vises Når bruker ser compact-variant Så vises innhold 1`] = `
<div>
  <div
    class="notification-panel notification-panel--info notification-panel__compact notification-panel__compact--basic notification-panel--has-children"
    data-analyticsid="notification-panel"
  >
    <span
      class="notification-panel__icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="#08667C"
        focusable="false"
        height="38"
        role="presentation"
        style="min-width: 38px; min-height: 38px;"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M26.367 31.224h-4.7v-.543a8.88 8.88 0 00.38-.037c.124-.015.23-.04.32-.075a.612.612 0 00.334-.255.881.881 0 00.097-.443v-6.324a.868.868 0 00-.12-.458 1.001 1.001 0 00-.297-.315 1.543 1.543 0 00-.405-.171 2.644 2.644 0 00-.495-.105v-.543l3.645-.192.11.11v7.9c0 .174.038.322.112.443a.67.67 0 00.32.271 2.363 2.363 0 00.699.194v.543zm-2.64-14.998a1.556 1.556 0 11-.002 3.111 1.556 1.556 0 01.002-3.11zM24 8.728C15.58 8.728 8.728 15.578 8.728 24c0 8.42 6.852 15.272 15.273 15.272 8.42 0 15.27-6.852 15.27-15.272 0-8.421-6.85-15.272-15.27-15.272z"
        />
      </svg>
    </span>
    <div
      class="notification-panel__content"
      id="test"
    />
  </div>
  <div
    class="notification-panel notification-panel--info notification-panel__compact notification-panel__compact--outline notification-panel--has-children"
    data-analyticsid="notification-panel"
  >
    <span
      class="notification-panel__icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="#08667C"
        focusable="false"
        height="38"
        role="presentation"
        style="min-width: 38px; min-height: 38px;"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M26.367 31.224h-4.7v-.543a8.88 8.88 0 00.38-.037c.124-.015.23-.04.32-.075a.612.612 0 00.334-.255.881.881 0 00.097-.443v-6.324a.868.868 0 00-.12-.458 1.001 1.001 0 00-.297-.315 1.543 1.543 0 00-.405-.171 2.644 2.644 0 00-.495-.105v-.543l3.645-.192.11.11v7.9c0 .174.038.322.112.443a.67.67 0 00.32.271 2.363 2.363 0 00.699.194v.543zm-2.64-14.998a1.556 1.556 0 11-.002 3.111 1.556 1.556 0 01.002-3.11zM24 8.728C15.58 8.728 8.728 15.578 8.728 24c0 8.42 6.852 15.272 15.273 15.272 8.42 0 15.27-6.852 15.27-15.272 0-8.421-6.85-15.272-15.27-15.272z"
        />
      </svg>
    </span>
    <div
      class="notification-panel__content"
      id="test"
    />
  </div>
</div>
`;

exports[`Gitt at NotificationPanel skal vises Når panelet kan lukkes Så kalles onClick-handleren når man klikker på knappen 1`] = `
<div>
  <div
    class="notification-panel notification-panel--info notification-panel--has-children notification-panel--dismissable"
    data-analyticsid="notification-panel"
  >
    <span
      class="notification-panel__icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="#08667C"
        focusable="false"
        height="48"
        role="presentation"
        style="min-width: 48px; min-height: 48px;"
        viewBox="0 0 48 48"
        width="48"
      >
        <path
          d="M26.372 31.239h-4.709v-.544c.129-.01.256-.022.38-.037a1.31 1.31 0 00.32-.075.614.614 0 00.335-.256.883.883 0 00.097-.444v-6.337a.87.87 0 00-.119-.459 1.003 1.003 0 00-.298-.316 1.546 1.546 0 00-.406-.171 2.65 2.65 0 00-.496-.105v-.544l3.652-.193.111.111v7.917c0 .174.038.322.112.443a.672.672 0 00.321.272 2.368 2.368 0 00.7.194v.544zM23.726 16.21a1.559 1.559 0 11-.002 3.118 1.559 1.559 0 01.002-3.118zm.275-7.514C15.562 8.696 8.696 15.561 8.696 24c0 8.438 6.866 15.304 15.305 15.304 8.438 0 15.303-6.866 15.303-15.304 0-8.439-6.865-15.304-15.303-15.304z"
        />
      </svg>
    </span>
    <span
      class="notification-panel__close"
    >
      <button
        aria-label="Lukk"
        class="close"
        data-analyticsid="close"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="hnds-style-icon"
          data-analyticsid="icon"
          fill="#000000"
          focusable="false"
          height="48"
          role="presentation"
          style="min-width: 48px; min-height: 48px;"
          viewBox="0 0 48 48"
          width="48"
        >
          <path
            d="M25.403 24l10.259-10.259-1.403-1.403L24 22.597l-10.259-10.26-1.403 1.403L22.597 24 12.338 34.26l1.403 1.403L24 25.403l10.259 10.259 1.403-1.403z"
          />
        </svg>
      </button>
    </span>
    <div
      class="notification-panel__content"
      id="test"
    >
      <div
        class="notification-panel__children"
      >
        Some text here for testing.
      </div>
    </div>
  </div>
</div>
`;

exports[`Gitt at NotificationPanel skal vises Når panelet vises med default props Så rendres komponenten riktig 1`] = `
<div>
  <div
    class="notification-panel notification-panel--info notification-panel--has-children"
    data-analyticsid="notification-panel"
  >
    <span
      class="notification-panel__icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="#08667C"
        focusable="false"
        height="48"
        role="presentation"
        style="min-width: 48px; min-height: 48px;"
        viewBox="0 0 48 48"
        width="48"
      >
        <path
          d="M26.372 31.239h-4.709v-.544c.129-.01.256-.022.38-.037a1.31 1.31 0 00.32-.075.614.614 0 00.335-.256.883.883 0 00.097-.444v-6.337a.87.87 0 00-.119-.459 1.003 1.003 0 00-.298-.316 1.546 1.546 0 00-.406-.171 2.65 2.65 0 00-.496-.105v-.544l3.652-.193.111.111v7.917c0 .174.038.322.112.443a.672.672 0 00.321.272 2.368 2.368 0 00.7.194v.544zM23.726 16.21a1.559 1.559 0 11-.002 3.118 1.559 1.559 0 01.002-3.118zm.275-7.514C15.562 8.696 8.696 15.561 8.696 24c0 8.438 6.866 15.304 15.305 15.304 8.438 0 15.303-6.866 15.303-15.304 0-8.439-6.865-15.304-15.303-15.304z"
        />
      </svg>
    </span>
    <div
      class="notification-panel__content"
      id="test"
    >
      <div
        class="notification-panel__children"
      >
        Some text here for testing.
      </div>
    </div>
  </div>
</div>
`;
